service/src/main/java/bio/terra/workspace/service/grant/flight/RevokeTemporaryGrantFlight.java

```diff
@@ -0,0 +1,439 @@
+      Boolean skip = context.getWorkingMap().get(SKIP, Boolean.class);
+      if (skip == null || skip.equals(TRUE)) {
+        logger.info("Skipping delete grant");
+      } else {
+        grantDao.deleteGrant(grantId, context.getFlightId());
+      }
+      return StepResult.getStepResultSuccess();
+    }
+
+    @Override
+    public StepResult undoStep(FlightContext context) throws InterruptedException {
+      // No undo possible if the delete fails. Dismal failure. Requires a human
+      // to look at the DB and see what is up with the grant.
+      return new StepResult(
+          StepStatus.STEP_RESULT_FAILURE_FATAL,
+          new InternalLogicException("Possible corruption of grant id" + grantId));
+    }
+  }
+
+  /**
+   * Step 2: revoke the permission A step that illustrates the consistent approach to IAM in GCP
+   * (not!)
+   */
+  public static class RevokeStep implements Step {
+    public static Logger logger = LoggerFactory.getLogger(RevokeStep.class);
+    private final GcpCloudContextService gcpCloudContextService;
+    private final CrlService crlService;
+    private final GrantDao grantDao;
+    private final ControlledResourceService controlledResourceService;
+    private final UUID grantId;
+
+    public RevokeStep(
+        GcpCloudContextService gcpCloudContextService,
+        CrlService crlService,
+        GrantDao grantDao,
+        ControlledResourceService controlledResourceService,
+        UUID grantId) {
+      this.gcpCloudContextService = gcpCloudContextService;
+      this.crlService = crlService;
+      this.grantDao = grantDao;
+      this.controlledResourceService = controlledResourceService;
+      this.grantId = grantId;
+    }
+
+    @Override
+    public StepResult doStep(FlightContext context) throws InterruptedException, RetryException {
+      Boolean skip = context.getWorkingMap().get(SKIP, Boolean.class);
+      if (skip == null || skip.equals(TRUE)) {
+        logger.info("Skipping revoke");
+        return StepResult.getStepResultSuccess();
+      }
+
+      // Get the grant data - we locked it, so if it is not there something is wrong
+      GrantData grantData = grantDao.getGrant(grantId);
+      if (grantData == null) {
+        throw new InternalLogicException("Locked grant not found: " + grantId);
+      }
+
+      try {
+        logger.info("Revoking grant {} type {}", grantData.grantId(), grantData.grantType());
+        switch (grantData.grantType()) {
+          case RESOURCE -> revokeResource(grantData);
+          case PROJECT -> revokeProject(grantData);
+          case ACT_AS -> revokeActAs(grantData);
+        }
+      } catch (ConflictException | IOException e) {
+        return new StepResult(StepStatus.STEP_RESULT_FAILURE_RETRY, e);
+      }
+      return StepResult.getStepResultSuccess();
+    }
+
+    @Override
+    public StepResult undoStep(FlightContext context) throws InterruptedException {
+      return StepResult.getStepResultSuccess();
```

Comment:
For 1, I'd agree that retrying is better than not, though I also don't have an easy way to spot perpetual failures (or at least one that's better than "keep a close eye on logs for a bit as this releases"

I think it's technically possible for this step to fail after the revoke happens (e.g. Stairway fails to write the "step finished" status to DB for some reason). In that case you're right though, this would undo (per case 1 above) and then succeed on the next run because this is robust to the GCP binding being gone already. So, agreed on not undoing here!